src/Workspaces/Remote/ServiceHub.CoreComponents/Microsoft.CodeAnalysis.Remote.ServiceHub.CoreComponents.csproj

```diff
@@ -20,27 +22,121 @@
+    <ItemGroup>
+      <!-- Find all Roslyn assemblies that we want to crossgen -->
+      <_R2RAssemblies Include="@(ReferencePath->'%(FileName)%(Extension)')" Condition="'%(ReferenceSourceTarget)' == 'ProjectReference'" />
+
+      <!-- Also crossgen a few assemblies we don't own but insert into VS -->
+      <_R2RAssemblies Include="Microsoft.ServiceHub.Framework.dll" />
+      <_R2RAssemblies Include="Microsoft.VisualStudio.Composition.dll" />
+      <_R2RAssemblies Include="Microsoft.VisualStudio.Telemetry.dll" />
+      <_R2RAssemblies Include="Microsoft.VisualStudio.Threading.dll" />
+      <_R2RAssemblies Include="MessagePack.dll" />
+      <_R2RAssemblies Include="Nerdback.Streams.dll" />
+      <_R2RAssemblies Include="Newtonsoft.Json.dll" />
+      <_R2RAssemblies Include="SQLitePCLRaw.batteries_v2.dll" />
+      <_R2RAssemblies Include="StreamJsonRpc.dll" />
+
+      <!-- Find all assemblies (including Roslyn and all dependencies) from the actual published location -->
+      <_AllPublishedAssemblyPaths Include="$(PublishDir)\**\*.dll" Exclude="$(PublishDir)\**\*.resources.dll" />
+      <_AllPublishedAssemblies Include="@(_AllPublishedAssemblyPaths->'%(FileName)%(Extension)')" >
+        <_FullFilePath>%(FullPath)</_FullFilePath>
+      </_AllPublishedAssemblies>
+
+      <!-- Get all non-Roslyn assemblies as a subset of all published ones, we don't crossgen those -->
+      <_NoR2RAssemblies Include="@(_AllPublishedAssemblies)" Exclude="@(_R2RAssemblies)" >
+        <_FullFilePath>%(_AllPublishedAssemblies._FullFilePath)</_FullFilePath>
+      </_NoR2RAssemblies>
+      <_NoR2RAssemblyPaths Include="@(_NoR2RAssemblies->'%(_FullFilePath)')" />
+
+      <!-- Now we get all Roslyn assemblies in the publish folder -->
+      <_R2RAssemblyPaths Include="@(_AllPublishedAssemblyPaths)" Exclude="@(_NoR2RAssemblyPaths)" />
+    </ItemGroup>
+    
+    <PropertyGroup>
+      <CrossgenWorkDir>$(PublishDir)\CrossGen\</CrossgenWorkDir>
+      <OriginalAssemblyFolder>$(CrossgenWorkDir)OriginalAssemblies\</OriginalAssemblyFolder>
+    </PropertyGroup>   
+  </Target>
+
+  <Target Name="PrepareCrossgenTargets" DependsOnTargets="LocateCrossgenTargets" Condition="'$(GenerateReadyToRun)' == 'true'">
+    <Move SourceFiles="@(_R2RAssemblyPaths)" DestinationFolder="$(OriginalAssemblyFolder)">
+      <Output TaskParameter="DestinationFiles" ItemName="_FilesWritten" />
+    </Move>
+    <ItemGroup>
+      <_CrossgenTargetPaths Include="$(OriginalAssemblyFolder)\*.dll" />
+      <_CrossgenTargetsAsDependencies Include="@(_CrossgenTargetPaths)" />
+      <_NonCrossgenTargetsAsDependencies Include="@(_NonRuntimeAssembliesInPublishFolder)" Exclude="@(_R2RAssemblyPaths)" />
+    </ItemGroup>
+  </Target>
+
+  <Target Name="CompileReadyToRun" DependsOnTargets="LocateDependencies;PrepareCrossgenTargets"  Condition="'$(GenerateReadyToRun)' == 'true'" Outputs="%(_CrossgenTargetPaths.FullPath)">    
```

Comment:
>_CrossgenTargetPaths [](http://example.com/codeflow?start=154&length=20)

Aren't these inputs?

Reply:
I read it somewhere that this is how to loop over an item group, basically something equivalent to 
```
foreach(var item in @(_CrossgenTargetPaths))
{
    CompileReadyToRun(item)
}
```

So I guess this is not really how it's supposed to be done?